File: 78-Subsets/solution.java

public class Solution {
    public List<List<Integer>> subsets(int[] nums) {
		List<List<Integer>> res = new ArrayList<>();
		List<Integer> subres = new ArrayList<>();

		if(nums.length == 0 || nums == null)	return res;

		Arrays.sort(nums);
		helper(res, subres, nums, 0);
		return res;
    }

    private void helper(List<List<Integer>> res, List<Integer> subres, int[] nums, int pos){
    	res.add(new ArrayList<Integer>(subres));

    	for(int i = pos; i < nums.length; i++) {
    		subres.add(nums[i]);
    		helper(res, subres, nums, i + 1);
    		subres.remove(subres.size() - 1);
    	}
    }
}

File: 78-Subsets/solution.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        lis = []
        res = []
        nums.sort()
        if nums is None or len(nums) == 0:
            return res;
        self.helper(res, lis, nums, 0)
        return res
        
    def helper(self, res, lis, nums, pos):
        res.append(list(lis))
        
        for i in xrange(pos, len(nums)):
            lis.append(nums[i])
            self.helper(res, lis, nums, pos + 1)
            lis.pop()
    
    
            